File: 2024/day_10/solution.js
function compile(instructions) {
  const registers = {};
  let pointer = 0;

  while (pointer < instructions.length) {
    const [op, arg1, arg2] = instructions[pointer].split(" ");

    switch (op) {
      case "MOV":
        registers[arg2] = isNaN(arg1) ? registers[arg1] || 0 : Number(arg1);
        break;
      case "INC":
        registers[arg1] = (registers[arg1] || 0) + 1;
        break;
      case "DEC":
        registers[arg1] = (registers[arg1] || 0) - 1;
        break;
      case "JMP":
        if ((registers[arg1] || 0) === 0) {
          pointer = Number(arg2);
          continue;
        }

        break;
    }

    pointer++;
  }

  return registers["A"];
}

const instructions = ["MOV -1 C", "INC C", "JMP C 1", "MOV C A", "INC A"];

console.log(compile(instructions)); // -> 2

File: 2024/day_10/solution.py
def compile(instructions):
  registers = {}
  pointer = 0

  while pointer < len(instructions):
    parts = instructions[pointer].split(" ")
    op = parts[0]
    arg1 = parts[1]
    arg2 = parts[2] if len(parts) > 2 else None

    if op == "MOV":
      registers[arg2] = registers.get(arg1, 0) if arg1.isalpha() else int(arg1)
    elif op == "INC":
      registers[arg1] = registers.get(arg1, 0) + 1
    elif op == "DEC":
      registers[arg1] = registers.get(arg1, 0) - 1
    elif op == "JMP":
      if registers.get(arg1, 0) == 0:
        pointer = int(arg2)
        continue

    pointer += 1

  return registers.get("A")
